girder/models/model_base.py

```diff
@@ -569,32 +569,103 @@ def subtreeCount(self, doc):
+    """
+    Overwrite the projection filter to either include (in the case of an inclusion filter) or
+    not exclude (in the case of an exclusion filter) the contents of overwrite.
+
+    :param fields: A mask for filtering result documents by key, or None to return the full
+        document, passed to MongoDB find() as the `projection` param.
+    :type fields: `str, list of strings, or tuple of strings for fields to be included from the
+        document, or dict for an inclusion or exclusion projection`.
+    :param overwrite: Additional document key(s) to be included or not excluded in fields.
+    :type overwrite: `str, list of strings, or tuple of strings for fields to be included from the
+        document.`
+    """
+    if fields is None:
+        return fields
+
+    overwrite = list(overwrite)
+    if _isInclusionProjection(fields):
         if not isinstance(fields, dict):
-            # If this is a list/tuple/set, that means inclusion
-            return True
+            # Inclusion projection (str, list, or tuple)
+            copy = list(fields)
+            for entry in overwrite:
+                if entry not in copy:
+                    copy.append(entry)
+        else:
+            # Inclusion projection (dict)
+            copy = dict(fields)
+            for entry in overwrite:
+                copy[entry] = True
+    else:
+        # Exclusion projection (dict)
+        copy = dict(fields)
+        for entry in overwrite:
+            if entry in copy.keys():
+                del copy[entry]
+    return copy
+
+
+def _removeOverwrittenFields(doc, fields):
+    """
+    Edit the document to be consistent with what the user originally requested, undoing what may
+        have been overwritten by _overwriteFields().
+
+    :param doc: A document returned by MongoDB find()
+    :type doc: dict
+    :param fields: The origian mask for filtering result documents by key, as specified by the user.
+    :type fields: `str, list of strings, or tuple of strings for fields to be included from the
+        document, or dict for an inclusion or exclusion projection`.
+    """
+    if fields is None:
+        return
 
+    whitelist = []
+    if isinstance(fields, dict):
         for k, v in six.viewitems(fields):
-            if k != '_id':
-                # We are only allowed either inclusion or exclusion keys in a dict, there can be no
-                # mixing of these, with the only exception being that the `_id` key can be set as
-                # an exclusion field in a dict that otherwise holds inclusion fields.
-                return v
-
-        # Empty dict or just _id field
-        return fields.get('_id', True)
+            if not v:
+                doc.pop(k, None)
+            else:
+                whitelist.append(k)
+        if whitelist:
+            for k in list(doc.keys()):
+                if k not in whitelist:
+                    del doc[k]
+    else:
+        fields = list(fields)
```

Comment:
No need to make a copy of `fields` here (iterating through a `tuple` works the same, anyway). If it's a string, we would need to cast it to a list.

Also, let's add a test for the case when `fields` is a string.